xds/server.go

```diff
@@ -233,6 +251,116 @@ func (s *GRPCServer) newListenerWrapper(lis net.Listener) (*listenerWrapper, err
+		return
+	}
+
+	// If we got all the way here, it means the received update was a good one.
+	update.goodUpdate.Fire()
+}
+
+func (s *GRPCServer) handleSecurityConfig(config *xdsclient.SecurityConfig, lw *listenerWrapper) error {
+	// If xdsCredentials are not in use, i.e, the user did not want to get
+	// security configuration from the control plane, we should not be acting on
+	// the received security config here. Doing so poses a security threat.
+	if !s.xdsCredsInUse {
+		return nil
+	}
+
+	// Security config being nil is a valid case where the control plane has
+	// not sent any security configuration. The xdsCredentials implementation
+	// handles this by delegating to its fallback credentials.
+	if config == nil {
+		// We need to explicitly set the fields to nil here since this might be
+		// a case of switching from a good security configuration to an empty
+		// one where fallback credentials are to be used.
+		lw.xdsHI.SetRootCertProvider(nil)
+		lw.xdsHI.SetIdentityCertProvider(nil)
+		lw.xdsHI.SetRequireClientCert(false)
+		return nil
+	}
+
+	cpc := s.xdsC.BootstrapConfig().CertProviderConfigs
+	// Identity provider is mandatory on the server side.
+	identityProvider, err := buildProvider(cpc, config.IdentityInstanceName, config.IdentityCertName, true, false)
+	if err != nil {
+		return err
+	}
+
+	// A root provider is required only when doing mTLS.
+	var rootProvider certprovider.Provider
+	if config.RootInstanceName != "" {
+		rootProvider, err = buildProvider(cpc, config.RootInstanceName, config.RootCertName, false, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Close the old providers and cache the new ones.
+	if lw.cachedIdentity != nil {
+		lw.cachedIdentity.Close()
+	}
+	if lw.cachedRoot != nil {
+		lw.cachedRoot.Close()
+	}
+	lw.cachedRoot = rootProvider
+	lw.cachedIdentity = identityProvider
+
+	// We set all fields here, even if some of them are nil, since they
+	// could have been non-nil earlier.
+	lw.xdsHI.SetRootCertProvider(rootProvider)
+	lw.xdsHI.SetIdentityCertProvider(identityProvider)
+	lw.xdsHI.SetRequireClientCert(config.RequireClientCert)
+	return nil
+}
+
+func buildProviderFunc(configs map[string]*certprovider.BuildableConfig, instanceName, certName string, wantIdentity, wantRoot bool) (certprovider.Provider, error) {
+	cfg, ok := configs[instanceName]
+	if !ok {
+		return nil, fmt.Errorf("certificate provider instance %q not found in bootstrap file", instanceName)
+	}
+	provider, err := cfg.Build(certprovider.BuildOptions{
+		CertName:     certName,
+		WantIdentity: wantIdentity,
+		WantRoot:     wantRoot,
+	})
+	if err != nil {
+		// This error is not expected since the bootstrap process parses the
+		// config and makes sure that it is acceptable to the plugin. Still, it
+		// is possible that the plugin parses the config successfully, but its
+		// Build() method errors out.
+		return nil, fmt.Errorf("xds: failed to get security plugin instance (%+v): %v", cfg, err)
```

Comment:
Is this `xds` going to be redundant?

Reply:
The convention seems to be that we add the component prefix to errors that we generate. The same or closely related prefix might also get added by the prefix logger if the returned error is logged at some point. I still haven't looked at any real logs to see how useful/bad these prefixes are getting.

Reply:
This isn't a convention we should have, *unless* the error is leaving the xds component.  Let's be careful about these so our logs are clear and not cluttered.

Reply:
Can I leave it as-is for now? We should probably discuss this and come up with a convention that we can follow.

Reply:
I've removed the xds prefix from the errors returned from here.

I didn't mean to pushback on the change you are suggesting here. I'm only saying that this is widespread and that we should talk about it and decide how we want to go about structuring error messages.